geode-management/src/main/java/org/apache/geode/management/client/ClusterManagementServiceProvider.java

```diff
@@ -0,0 +1,135 @@
+ * can create an instance of the {@code ClusterManagementService} (CMS for short) in several ways,
+ * each providing various level of control or expediency.
+ * <p/>
+ * Calling {@code getFactory(context)} will return an explicit instance of {@link
+ * ClusterManagementServiceProviderFactory}. Methods on this factory can then be called to create or
+ * retrieve instances of a CMS. New {@link ClusterManagementServiceProvider}s can be written if
+ * specific customization or parameterization is required.
+ * <p/>
+ * A note about contexts. A context is simply a unique string which identifies a specific instance
+ * of CMS that will be returned. Contexts map to the different uses of a CMS. Currently, the
+ * following contexts are provided:
+ * <ul>
+ * <li>
+ * {@code JAVA_CLIENT_CONTEXT} ("java-client") - would be used to retrieve a CMS instance on a pure
+ * Java client - i.e. an app that is not running in the context of a Geode client or server. This
+ * context is available when using the <i>geode-management</i> module.
+ * </li>
+ * <li>{@code GEODE_CLIENT_CONTEXT} ("geode-client") - would be used to retrieve a CMS instance
+ * from a JVM where a {@code ClientCache} exists. This context is available when using the
+ * <i>geode-core</i> module.
+ * </li>
+ * <li>
+ * {@code LOCATOR_CONTEXT} ("locator") - would be used to retrieve a CMS instance from a JVM which
+ * is hosting a locator. This context should typically only be used for internal Geode code and is
+ * not particularly relevant to client or user applications.
+ * </li>
+ * <li>
+ * {@code SERVER_CONTEXT} ("server") - would be used to retrieve a CMS instance from a JVM where a
+ * {@code Cache} exists. This would be either a server or locator. This context is available when
+ * using the <i>geode-core</i> module.
+ * </li>
+ * </ul>
+ * If the URL of the Cluster Management Service is known, the {@code getService(url)} method can be
+ * called. This implicitly uses the {@code JAVA_CLIENT_CONTEXT}. For further control the CMS can be
+ * configured with a {@link ClientHttpRequestFactory} by calling {@code
+ * getService(requestFactory)}.
+ * <p/>
+ * Finally, the simplest way to create a CMS instance is simply to call {@code getService()}. This
+ * method
+ * will attempt to infer the context and use an appropriate service provider to create a CMS
+ * instance.
+ */
+@Experimental
+public class ClusterManagementServiceProvider {
+
+  public static final String JAVA_CLIENT_CONTEXT = "java-client";
+  public static final String GEODE_CLIENT_CONTEXT = "geode-client";
+  public static final String LOCATOR_CONTEXT = "locator";
+  public static final String SERVER_CONTEXT = "server";
+
+  private static List<ClusterManagementServiceProviderFactory> providerFactories = null;
+
+  public static ClusterManagementService getService() {
+    for (String context : new String[] {SERVER_CONTEXT, GEODE_CLIENT_CONTEXT,
```

Comment:
this list should only be SERVER_CONTEXT, LOCATOR_CONTEXT and GEODE_CLIENT_CONTEXT, because only those context support/will support create() method without parameter.